### PR TITLE
CRIMRE-248 update application review status to ready for assessment

### PR DESCRIPTION
**Pull request by hiboabd**
## Description of change

- Changes the review_state of the application in the Datastore to ready_for_assessment
- Action can only be completed once and an application can only transition from application_received to ready_for_assessment. An error is raised for all other statuses

## Link to relevant ticket
[CRIMRE-248](https://dsdmoj.atlassian.net/jira/software/c/projects/CRIMRE/boards/960?modal=detail&selectedIssue=CRIMRE-248)

## Notes for reviewer / how to test

Me and Tim discussed the investigation to add an index to enforce uniqueness of USN. I have left a comment in the [ticket](https://dsdmoj.atlassian.net/jira/software/c/projects/CRIMRE/boards/960?modal=detail&selectedIssue=CRIMRE-248) explaining how this could be tackled

[CRIMRE-248]: https://dsdmoj.atlassian.net/browse/CRIMRE-248?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/services/operations/mark_as_ready_application.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module Operations
+  class MarkAsReadyApplication
+    def initialize(application_id:)
+      @application = CrimeApplication.find(application_id)
+    end
+
+    def call
+      raise Errors::AlreadyMarkedAsReady if application.ready_for_assessment?
+      raise Errors::AlreadyCompleted if application.assessment_completed?
+      raise Errors::AlreadyReturned if application.returned?
```
Comment (hiboabd) [inline app/services/operations/mark_as_ready_application.rb:10]: I considered simplifying these errors into one e.g. CanOnlyBeRecieved but I wondered if the ability to differentiate between which status raised the error was going to be helpful down the line.. but open to opinions

Comment (arthurashman) [reply app/services/operations/mark_as_ready_application.rb:10]: I think this is good. Probably useful to know what state something was in when the user tried to mark as ready for assessment - would make it easier to debug if ever it did happen

File: spec/api/datastore/v2/reviewing/complete_application_spec.rb [hunk 2]
```diff
@@ -31,21 +31,15 @@
         application.update!(review_status: :assessment_completed, reviewed_at: 1.week.ago)
       end
 
-      it 'raises a 409 error' do
-        api_request
-        expect(response).to have_http_status :conflict
-      end
+      it_behaves_like 'raises a 409 error'
```
Comment (arthurashman) [inline spec/api/datastore/v2/reviewing/complete_application_spec.rb:34]: Nice

Comment (arthurashman) [reply spec/api/datastore/v2/reviewing/complete_application_spec.rb:34]: Wonder if we can reword this to a sentence` it_behaves_like 'raises a 409 error'` e.g. ` it_behaves_like 'an error that raises a 409 status code'`

File: spec/api/datastore/v2/reviewing/mark_as_ready_application_spec.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  describe 'PUT /api/applications/application_id/mark_as_ready' do
+    subject(:api_request) do
+      put(
+        "/api/v2/applications/#{application.id}/mark_as_ready"
+      )
+    end
+
+    context 'with a submitted application' do
+      it 'marks the application as ready for' do
```
Comment (arthurashman) [inline spec/api/datastore/v2/reviewing/mark_as_ready_application_spec.rb:18]: `      it 'marks the application as ready for assessment' do`

**Review by arthurashman — APPROVED** [review]
Just some small non-blocking comments to do with naming. Otherwise looks great!
